src/Stratis.Bitcoin/Connection/ConnectionManager.cs

```diff
@@ -123,7 +120,6 @@ public IReadOnlyNetworkPeerCollection ConnectedPeers
             this.Parameters.UserAgent = $"{this.ConnectionSettings.Agent}:{versionProvider.GetVersion()} ({(int)this.NodeSettings.ProtocolVersion})";
 
             this.Parameters.Version = this.NodeSettings.ProtocolVersion;
-            this.connectedPeersQueue = new AsyncQueue<INetworkPeer>(this.OnPeerAdded);
```

Comment:
If you delete this you introduce back the race condition, the async queue processes single threaded all attempts to check the filtering.

Reply:
My point still stands that we can't check the range filtering after we have already added the peer to the `connectedPeers` collection. So currently we will always disconnect straight after we've added as the peer we are checking is already in the collection.

Reply:
>So currently we will always disconnect straight after we've added as the peer we are checking is already in the collection.

No, because we ignore current peer during checking the groups: 

```
if (peer == connectedPeer)
    continue;
```

Reply:
@noescape00 yes I wrongly assumed there :)